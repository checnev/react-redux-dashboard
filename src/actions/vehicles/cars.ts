import CarsService from 'services/CarsService';
import { push, CallHistoryMethodAction } from 'connected-react-router';

import type { RootState } from 'store';
import type { ThunkAction } from 'redux-thunk';
import type { Car } from 'reducers/vehicles/cars';

interface GetCarsStartAction {
  type: typeof GET_CARS_START;
}

interface GetCarsSuccessAction {
  type: typeof GET_CARS_SUCCESS;
  payload: Car[];
}

interface GetCarsFailAction {
  type: typeof GET_CARS_FAIL;
  payload: string;
}

export type CarsActionTypes = GetCarsStartAction 
  | GetCarsSuccessAction 
  | GetCarsFailAction
  | CallHistoryMethodAction;

export const GET_CARS_START = 'GET_CARS_START';
export const GET_CARS_SUCCESS = 'GET_CARS_SUCCESS';
export const GET_CARS_FAIL = 'GET_CARS_FAIL';

export const getCarById = (id: string): ThunkAction<void, RootState, unknown, CarsActionTypes> => (
  async (dispatch) => {
    dispatch({
      type: GET_CARS_START,
    });

    try {
      const car = await CarsService.getCarById(id);

      if (!car) throw new Error(`Car with id ${id} doesn't exist.`);

      dispatch({
        type: GET_CARS_SUCCESS,
        payload: [car],
      });
    } catch (err) {
      dispatch({
        type: GET_CARS_FAIL,
        payload: err.message,
      });

      dispatch(push('/404'));
    }
  }
);

export const getCars = (): ThunkAction<void, RootState, unknown, CarsActionTypes> => (
  async (dispatch) => {
    dispatch({
      type: GET_CARS_START,
    });

    try {
      const cars = await CarsService.getCars();
      dispatch({
        type: GET_CARS_SUCCESS,
        payload: cars,
      });
    } catch (err) {
      dispatch({
        type: GET_CARS_FAIL,
        payload: err.message,
      });
    }
  }
);
