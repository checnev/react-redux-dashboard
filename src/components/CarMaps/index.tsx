import React, { useRef } from 'react';
import { YMaps, Map, MapState, Placemark, YMapsApi } from 'react-yandex-maps';
import { Skeleton } from '@material-ui/lab';
import { STATUS_AS_STRING } from 'constants/car';

import type { Car } from 'reducers/vehicles/cars';


interface CarMapsProps {
  cars: Car[];
  width?: string | number;
  height?: string | number;
}

const CarMaps: React.FC<CarMapsProps> = React.memo(({
  cars,
  width = '100%',
  height = 250 
}: CarMapsProps) => {
  const map = useRef<any>(null);
  const initMap = React.useCallback((ymaps: YMapsApi) => {
    if (cars.length > 1) {
      map.current.setBounds(
        map.current.geoObjects.getBounds(),
        {
          checkZoomRange: true,
          zoomMargin: 15,
        }
      );
    }
  }, [cars]);
  const defaultState: MapState | null = React.useMemo(() => {
    if (!cars || !cars.length) {
      return null;
    }
    
    return {
      zoom: 15,
      center: cars[0].coordinats,
    };
  }, [cars]);

  return (
    <>
      {!cars.length && <Skeleton variant="rect" width={width} height={height} />}
      {cars.length > 0 && defaultState && (
        <YMaps>
          <Map
            width={width}
            height={height}
            defaultState={defaultState}
            instanceRef={(ref: any) => map.current = ref}
            onLoad={(ymaps: YMapsApi) => {
              ymaps.ready(initMap);
            }}
          >
            {cars.map((car) => <Placemark
              key={car.id}
              geometry={car.coordinats}
              properties={{
                iconCaption: car.model,
                balloonContentHeader: car.model,
                balloonContent: STATUS_AS_STRING[car.status],
              }}
              modules={['geoObject.addon.balloon']}
              options={{
                iconColor: car.color,
              }}
            />)}
          </Map>
        </YMaps>
      )}
    </>
  );
});

export default CarMaps;
