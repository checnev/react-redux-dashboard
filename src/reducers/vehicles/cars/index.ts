import {
  GET_CARS_START,
  GET_CARS_SUCCESS,
  GET_CARS_FAIL,
  CarsActionTypes,
} from 'actions/vehicles/cars';

export enum CarStatus {
  Free,
  Busy,
  Repair,
}

export interface Car {
  id: number;
  stateNumber: string;
  model: string;
  status: CarStatus;
  color: string;
  imageUrl: string;
  coordinats: number[];
}

export interface CarsState {
  cars: Car[];
  isFetching: boolean;
}

const initialState: CarsState  = {
  cars: [],
  isFetching: false,
};

const carsReducer = (
  state = initialState,
  action: CarsActionTypes
): CarsState => {
  switch (action.type) {
    case GET_CARS_START:
      return { ...state, isFetching: true };
    case GET_CARS_SUCCESS: 
      return {
        ...state,
        isFetching: false,
        cars: action.payload,
      };
    case GET_CARS_FAIL:
      return {
        ...state,
        isFetching: false,
        cars: [],
      };
    default:
      return { ...state };
  }
};

export default carsReducer;
