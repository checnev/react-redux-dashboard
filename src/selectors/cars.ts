import { createSelector } from 'reselect';
import type { RootState } from 'store';
import type { Car } from 'reducers/vehicles/cars';
import type { CarReport, CarReports } from 'reducers/vehicles/reports';

export interface CarWithReport extends Car {
  report?: CarReport;
}

export const carsSelector = (state: RootState) => state.vehicles.cars.cars;
export const reportsSelector = (state: RootState) => state.vehicles.reports.reports;

export const carWithReportsSelector = createSelector(
  carsSelector,
  reportsSelector,
  (cars: Car[], reports: CarReports | null) => {
    const carList: CarWithReport[] = [];

    for (let car of cars) {
      const carReports = reports?.get(car);
      const carData: CarWithReport = Object.assign({}, car);

      if (carReports) {
        carData.report = carReports.reduce((summary: CarReport, report) => {
          summary.income += report.income;
          summary.countOfLeases += report.countOfLeases;
          summary.incidents += report.incidents;

          return summary;
        }, {
          income: 0,
          countOfLeases: 0,
          incidents: 0,
        });
      }

      carList.push(carData);
    }

    return carList;
  }
);
