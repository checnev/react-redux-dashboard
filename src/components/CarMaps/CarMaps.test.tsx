import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import pretty from 'pretty';
import CarMaps from 'components/CarMaps';
import cars from 'test/data/cars';

jest.mock('react-yandex-maps', () => ({
  YMaps: (props: any) => <div data-testid="YMaps" {...props}></div>,
  Map: (props: any) => {
    const { defaultState, instanceRef, ...otherProps } = props;
    return <div data-testid="Map" {...otherProps}></div>
  },
  Placemark: (props: any) => <div data-testid="Placemark" {...props}></div>,
}));

let container: any = null;

beforeEach(() => {
  container = document.createElement('div');
  document.body.append(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('CarMaps render', () => {
  act(() => {
    render(<CarMaps cars={cars} />, container);
  });

  expect(
    pretty(container.innerHTML)
  ).toMatchSnapshot();
});
