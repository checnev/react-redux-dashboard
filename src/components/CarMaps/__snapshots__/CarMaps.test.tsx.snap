// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarMaps render 1`] = `
"<div data-testid=\\"YMaps\\">
  <div data-testid=\\"Map\\" width=\\"100%\\" height=\\"250\\">
    <div data-testid=\\"Placemark\\" geometry=\\"1,1\\" properties=\\"[object Object]\\" modules=\\"geoObject.addon.balloon\\" options=\\"[object Object]\\"></div>
    <div data-testid=\\"Placemark\\" geometry=\\"1,1\\" properties=\\"[object Object]\\" modules=\\"geoObject.addon.balloon\\" options=\\"[object Object]\\"></div>
  </div>
</div>"
`;
